### PR TITLE
Support OCaml 4.09

### LINKED ISSUE #1028: Support OCaml 4.09
Please support OCaml 4.09

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #1028 
The only API change comes from https://github.com/ocaml/ocaml/commit/ed74b5b

(I expected more changes in the files, maybe using symbolic links from 4.08 would be enough)

### CODE REVIEW

File: src/ocaml/typing/409/typecore.ml [hunk 1]
```diff
@@ -0,0 +1,5300 @@
...
+let type_class_arg_pattern cl_num val_env met_env l spat =
+  reset_pattern None false;
+  let nv = newvar () in
+  let pat = type_pat ~no_existentials:In_class_args (ref val_env) spat nv in
+  if has_variants pat then begin
+    Parmatch.pressure_variants val_env [pat];
+    iter_pattern finalize_variant pat
+  end;
+  List.iter (fun f -> f()) (get_ref pattern_force);
+  if is_optional l then unify_pat val_env pat (type_option (newvar ()));
```
Comment (gpetiot) [inline src/ocaml/typing/409/typecore.ml]: The only API change is `mk_fconv`

**Comment by trefis** [comment]
:heart_eyes: 

> maybe using symbolic links from 4.08 would be enough

Let's not :)

I haven't looked at anything yet, but can you also update the travis script so that we test 4.09.0 instead of/in addition to 4.08.1?

**Comment by avsm** [comment]
Works for me on 4.09 with a smoke test!

**Comment by hannesm** [comment]
for me, even after applying https://github.com/gpetiot/merlin/pull/1, I get the following error:
```
/usr/home/hannes/.opam/4.09.0/lib/ocaml/stdlib.cmi
seems to be compiled with a version of OCaml that is not
supported by Merlin.
```

**Comment by trefis** [comment]
The error message seems to indicate that either:
- you're not actually using the version from this PR
- at least `src/ocaml/typing/magic_numbers.ml` wasn't update; and that there is another problem.

**Comment by gpetiot** [comment]
I didn't update `src/ocaml/typing/magic_numbers.ml` indeed, should the additional value be `"Caml1999I026"` for 4.09? I don't know where these numbers come from.

**Comment by Octachron** [comment]
The cmi magic number for 4.09 is indeed `"Caml1999I026"`: https://github.com/ocaml/ocaml/blob/4.09/utils/config.mlp#L87 .

**Comment by hannesm** [comment]
I extended (updated and force-pushed) https://github.com/gpetiot/merlin/pull/1 which results in a merlin with 4.09 support for me

**Comment by madroach** [comment]
Please also update ``merlin.opam``. Thanks!

**Comment by hannesm** [comment]
in my experience, https://github.com/gpetiot/merlin/pull/1/commits/154323808a14adbb7556e44d6eefb85387b4aecc is needed as well to get merlin working with 4.09...

**Comment by gpetiot** [comment]
I did, sorry for the rookie mistakes, I'm not familiar with the Merlin codebase yet.

**Comment by hannesm** [comment]
anything open for this to get merged and released? I'm using this since a while for development with emacs and 4.09.0, it works fine for me.

**Comment by xuanruiqi** [comment]
I would strongly be in favor in merging and releasing this ASAP. OCaml 4.09.0 has been out for a while and Merlin support is rather essential.

**Comment by copy** [comment]
This seems to have an issue with float in format types: `Printf.printf "%f" 0.0` reports:

```
This expression has type CamlinternalFormatBasics.float_kind_conv but an expression was expected of type CamlinternalFormatBasics.float_conv = CamlinternalFormatBasics.float_flag_conv * CamlinternalFormatBasics.float_kind_conv
```

**Comment by bschommer** [comment]
Beside the format types issue, is there anything hindering a merge of this PR?

**Comment by anmonteiro** [comment]
With these new changes I can't build merlin locally. Here's the error I get:

```
# File "src/ocaml/preprocess/parser_raw.ml", line 5, characters 2-42:
# 5 |   MenhirLib.StaticVersion.require_20190626
#       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound value MenhirLib.StaticVersion.require_20190626
# Hint: Did you mean require_20190924?
```

**Comment by nojb** [comment]
It is due to a mismatch between your `menhir` version and that used by the author of the PR. Either:
1) use the same version of `menhir` as the author (`20190626`), or:
2) regenerate the `merlin` parser using the your (newer) `menhir`.

**Comment by gpetiot** [comment]
Replaced by #1055

**Pull request closed without merging**